### ISSUE TITLE
autograd.optim wrapper does not update state value

### ISSUE THREAD

**Issue by hbqjzj** [twitter/torch-autograd#121]
It seems that autograd.optim wrapper doesn't update "state" value. It outputs "states", but it is impossible to use it iteratively.

**Comment by ghostcow** [comment]
That's fine, it gives you the states table it uses so you can manually change things per weight tensor (if you wish), but you don't actually pass it back to any function.
It's saved inside the wrapper closure already.

Here's how to use the optim wrapper:

1. optim/init.lua calls wrap() exactly once on initialization (per optimization method):
```
for k, v in pairs(require 'optim') do
   opt[k] = wrap(v)
end

return opt
```

2. you call `autograd.optim.sgd(df,state,params)` ONCE to get the optimizing function:
```
local df = autograd(f, {optimize = true})
local state = {learningRate=1e-2}
local optimizer, states = autograd.optim.sgd(df, state, params)
```
3. you then use `optimizer` to update your weights at each iteration:
```
local grads, loss = optimizer(data, target)
```

*Example adapted from optim tests [here](https://github.com/twitter/torch-autograd/blob/master/test/test.lua#L1313) and [here](https://github.com/twitter/torch-autograd/blob/master/test/test.lua#L1360)

**Comment by hbqjzj** [comment]
The first test case works because the learning rate does not decay. However, in most cases we want the learning rate to decay, and that needs the number of iteration, which is stored in the variable states. Additionally, the moments matrices are also stored in the variable states. The call 
```lua
local grads, loss = optimizer(data, target)
```
will always assume the number of iteration is zero (or one) and moment matrices are empty in every iteration.

**Comment by eugenium** [comment]
Did anyone find a good solution to this without repeated calls to autograd.optim? It does seem the wrapper doesn't handle the state properly as hbqjzj pointed out.

**Comment by szagoruyko** [comment]
@eugenium there doesn't seem to be an issue. state is created once and then kept as a local variable in optimizer.

**Comment by eugenium** [comment]
Ah yea I see now. thanks.

**Comment by synchro--** [comment]
@ghostcow So, is there a complete working  example (like the mnist one) on how to use Optim?
